ProcessOptimizer/model_systems/color_pH.py

```diff
@@ -0,0 +1,339 @@
+    delb = Lab1['b'] - Lab2['b']
+    result = math.sqrt(delL * delL + dela * dela + delb * delb)
+    return result
+
+
+def delta_e_2000(Lab1, Lab2, verbose=False, test=False, formula='Rochester'):
+    """
+    Takes Lab values as a dictionary and outputs a DeltaE2000 calculation
+
+    Example Dictionarys:
+    Lab1 = {'L': 50.00, 'a': 2.6772, 'b': -79.7751}
+    Lab2 = {'L': 50.00, 'a': 0.00, 'b': -82.7485}
+
+    # verbose=True
+
+    Prints out all of the calculations that comes up with the deltaE2000.
+
+    # test=True
+
+    Returns all of the calculations that comes up with the deltaE2000 in the below order:
+    a1Prime, a2Prime, c1Prime, c2Prime, h1Prime, h2Prime, hBarPrime, g, t, sL, sC, sH, rT, DE2000
+    Rounded to 4 decimal places, as that was what the test data set was rounded to. 
+
+    formula kwarg can be 'Rochester' or 'Bruce'.
+
+    Rochester uses a different calculation for hPrime, h1Prime, h2Prime and hBarPrime than Bruce
+    Read the white paper by Gaurav Sharma, Wencheng Wu and Endul N. Dala (http://www2.ece.rochester.edu/~gsharma/ciede2000/ciede2000noteCRNA.pdf)
+    """
+
+    kL = 1.0
+    kC = 1.0
+    kH = 1.0
+    lBarPrime = 0.5 * (Lab1['L'] + Lab2['L'])
+    c1 = math.sqrt(Lab1['a'] * Lab1['a'] + Lab1['b'] * Lab1['b'])
+    c2 = math.sqrt(Lab2['a'] * Lab2['a'] + Lab2['b'] * Lab2['b'])
+    cBar = 0.5 * (c1 + c2)
+    cBar7 = cBar**7
+    g = 0.5 * (1.0 - math.sqrt(cBar7 / (cBar7 + 25**7)))  # 25**7 or 6103515625
+
+    a1Prime = Lab1['a'] * (1.0 + g)
+    a2Prime = Lab2['a'] * (1.0 + g)
+
+    c1Prime = math.sqrt(a1Prime * a1Prime + Lab1['b'] * Lab1['b'])
+    c2Prime = math.sqrt(a2Prime * a2Prime + Lab2['b'] * Lab2['b'])
+
+    cBarPrime = 0.5 * (c1Prime + c2Prime)
+
+    if formula == 'Rochester':
+        # Rochester hPrime
+        def hPrime(B_i_star, aiPrime):
+            if B_i_star == 0 and aiPrime == 0:
+                return 0
+            else:
+                hPrime = math.atan2(B_i_star, aiPrime) * 180 / math.pi
+                if hPrime < 0:
+                    hPrime += 360
+                return hPrime
+
+        # Gets h1 and h2 Primes
+        h1Prime = hPrime(Lab1['b'], a1Prime)
+        h2Prime = hPrime(Lab2['b'], a2Prime)
+
+        # START Rochester hBarprime
+        if abs(h1Prime - h2Prime) <= 180 and c1Prime * c2Prime != 0:
+            hBarPrime = (h1Prime + h2Prime) / 2
+
+        elif abs(h1Prime - h2Prime) > 180 and (h1Prime + h2Prime) < 360 and c1Prime * c2Prime != 0:
+            hBarPrime = (h1Prime + h2Prime + 360) / 2
+
+        elif abs(h1Prime - h2Prime) > 180 and (h1Prime + h2Prime) >= 360 and c1Prime * c1Prime != 0:
+            hBarPrime = (h1Prime + h2Prime - 360) / 2
+
+        elif (c1Prime * c2Prime) == 0:
+            hBarPrime = (h1Prime + h2Prime)
+        # END Rochester hBarprime
+
+    elif formula == 'Bruce':
+
+        # Bruces hPrimes
+        h1Prime = (math.atan2(Lab1['b'], a1Prime) * 180.0) / math.pi
+
+        if (h1Prime < 0.0):
```

Comment:
Ugh, please do this with the modulo operator `%` (I know it is not your code, I just couldn't help myself.)

Reply:
No comments... We can change it. I'm still really happy for the code, so we should just reference the source and then do the changes.

Reply:
No, let's not change code we have gotten from elsewhere unless it is broken. And this isn't broken, I am just being snobbish (and perhaps erring on the side of hard to read code).